src/ray/core_worker/task_event_buffer.cc

```diff
@@ -143,72 +136,85 @@ void TaskEventBufferImpl::FlushEvents(bool forced) {
       return;
     }
 
-    if (buffer_.size() == 0) {
+    // No data to send.
+    if (buffer_.empty() && send_buffer_.empty()) {
       return;
     }
 
-    task_events.reserve(
-        RayConfig::instance().task_events_max_num_task_events_in_buffer());
-    buffer_.swap(task_events);
-    next_idx_to_overwrite_ = 0;
-
-    num_profile_task_events_dropped = num_profile_task_events_dropped_;
-    num_profile_task_events_dropped_ = 0;
+    // Have new data and send buffer empty, add new data to the send buffer.
+    if (send_buffer_.empty() && !buffer_.empty()) {
+      buffer_.swap(send_buffer_);
```

Comment:
what's the reason why we introduce send_buffer? Is it because if we just iterate & erase buffer, it screws up the next_index?

Reply:
Yes, but now I think about it, maybe this is ok with iterating backward, i.e. sending the **most recently added** ones first (which simply back offs the next index or reducing the buffer size) 

Alternatively, we could send the least recently added ones, but erasing them will result in data movement or leaving holes in the buffer.